lib/puppet_x/puppetlabs/powershell_manager.rb

```diff
@@ -0,0 +1,244 @@
+require 'securerandom'
+require 'open3'
+require 'ffi'
+
+module PuppetX
+  module PowerShell
+    class PowerShellManager
+      extend Puppet::Util::Windows::String
+      extend FFI::Library
+
+      @@instances = {}
+
+      def self.instance(cmd)
+        @@instances[:cmd] ||= PowerShellManager.new(cmd)
+      end
+
+      def initialize(cmd)
+        @stdin, @stdout, @ps_process = Open3.popen2(cmd)
+
+        Puppet.debug "#{Time.now} #{cmd} is running as pid: #{@ps_process[:pid]}"
+
+        at_exit { exit }
+      end
+
+      def execute(powershell_code, timeout_ms = 300 * 1000)
+        output_ready_event_name =  "Global\\#{SecureRandom.uuid}"
+        output_ready_event = self.class.create_event(output_ready_event_name)
+
+        # always need a trailing newline to ensure PowerShell parses code
+        code = <<-CODE
+        $event = [Threading.EventWaitHandle]::OpenExisting("#{output_ready_event_name}")
+        if ($runspace -eq $null)
+        {
+          $runspace = [RunspaceFactory]::CreateRunspace()
+          $runspace.Open()
+        }
+
+        $powershell_code = @'
+#{powershell_code}
+'@
+        $ps = $null
+
+        try
+        {
+          # http://learn-powershell.net/2012/05/13/using-background-runspaces-instead-of-psjobs-for-better-performance/
+          $ps = [powershell]::create()
+          $ps.Runspace = $runspace
+          [Void]$ps.AddScript($powershell_code)
+
+          $asyncResult = $ps.BeginInvoke()
+
+          if (!$asyncResult.AsyncWaitHandle.WaitOne(#{timeout_ms}, $false))
+          {
+            throw "Catastrophic failure: PowerShell DSC resource timeout (#{timeout_ms} ms) exceeded while executing"
+          }
+
+          $output = $ps.EndInvoke($asyncResult)
+
+          if($ps.HadErrors){
```

Comment:
Catch non terminating errors. However, did we treat non-terminating as terminating before?